packages/vuelidate/src/core.js

```diff
@@ -147,20 +148,19 @@ function createAsyncResult (rule, model, initResult, $pending) {
  * @param {String} key
  * @return {{$params: *, $message: Ref<String>, $pending: Ref<Boolean>, $invalid: Ref<Boolean>}}
  */
-function createValidatorResult (rule, state, key) {
-  const model = computed(() => state[key].value)
-  const ruleResult = callRule(rule.$validator, model)
+function createValidatorResult (rule, state, key, $dirty) {
+  const model = computed(() => unwrap(unwrap(state)[key]))
 
   const $pending = ref(false)
   const $params = rule.$params || {}
-  const $invalid = isPromise(ruleResult)
+  const $invalid = rule.$async
```

Comment:
Is this the new convention where we ask them to apply this to async validators?

What do we do with backwards compat? We need to have some sort of fallback or detection, otherwise the validator will always return truthy, the `promise` is an object :)

Maybe we can check if after calling this in the computed, it returns a promise, and if it does, we throw or warn?

Reply:
Yeah! I figured that now that we don’t do validations upfront until the state is dirty, we have no easy way to assign different handlers for validators. To keep it consistent, I figured having a `$async` flag makes it really easy to navigate the different strategy. I think that depending on user feedback we could then also implement some other ways to handle async validations. For example maybe validations based on status codes? Something to think about.

But I agree with the fallback detection. Throwing an error for the developer there would make perfect sense!

Reply:
I was also thinking if we could make all validators `async` by default. This will complicate it though...
But we can ofc just throw. Should we do this here, or file a separate issue to track it later? 

Also we can offer a helper, like Vue does for async functional components. Similar to `withMessage` and `withProps`, `email: { isTaken: withAsync(asyncCheck) }`

Reply:
Let’s have it as a separate PR.